nixos/munin: New options (and some bugfixes) for service configuration

ToxicFrog opened the pull request:
###### Motivation for this change

The old Munin configuration had some outright bugs (wrong fonts used for graph generation, documentation links out of date) and was missing a lot of useful knobs. These changes fix those bugs, and add options for plugin configuration, loading extra plugins, disabling builtin plugins, and applying custom styles to the HTML that Munin generates.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

ToxicFrog committed:
nixos/munin: fix documentation links

Since this module was written, Munin has moved their documentation from
munin-monitoring.org/wiki to guide.munin-monitoring.org. Most of the
links were broken, and the ones that weren't went to "please use the new
site" pages.

ToxicFrog committed:
nixos/munin: add extraPluginConfig option

This lets you specify additional plugin-specific configuration to go in
plugin-conf.d, and complements the extraConfig and extraGlobalConfig
options.

ToxicFrog committed:
nixos/munin: add disabledPlugins option

This is just a set of globs to remove from the active plugins directory
after autoconfiguration is complete.

I also removed the hard-coded disabling of "diskstats", since it seems
to work just fine now.

ToxicFrog committed:
nixos/munin: require DejaVu fonts if enabled

munin-graph is hardcoded to use DejaVu Mono for the graph legends; if it
can't find it, there's no guarantee it finds a monospaced font at all,
and if it can't find a monospaced font the legends come out badly
misformatted.

ToxicFrog committed:
nixos/munin: add extraPlugins and extraAutoPlugins options [#17895]

extraAutoPlugins lets you list plugins and plugin directories to be
autoconfigured, and extraPlugins lets you enable plugins on a one-by-one
basis. This can be used to enable plugins from contrib (although you'll
need to download and check out contrib yourself, then point these
options at it), or plugins you've written yourself.

ToxicFrog committed:
nixos/munin: add extraCSS option

This permits custom styling of the generated HTML without needing to
build your own Munin package from source. Also comes with an example
that works as a passable dark theme for Munin.

ToxicFrog commented:
On further testing, the extraAutoPlugins option doesn't work reliably due to the way Munin searches for plugins. I'll see about fixing that this evening.

florianjacob commented on nixos/modules/services/monitoring/munin.nix:
```diff
@@ -24,6 +24,10 @@ let
       logdir    /var/log/munin
       rundir    /run/munin
 
+      ${if cronCfg.extraCSS != ""
```
How about
```
      ${lib.optionalString (cronCfg.extraCSS != "") "staticdir ${customStaticDir"}}
```
- should give the same result and be more explicit.

ToxicFrog replied:
Done.

florianjacob commented:
`nixos/tests/munin.nix` disables a plugin via `ignore_file ^apc_nis$`, it would make sense to switch that over to the new module option. Regarding previous `ignore_file` way vs. deleting the plugins like your option does, will that make any difference for munin? There is no other way for plugins to be enabled than linking them to `/etc/munin/plugins`, right?

Overall, this gives the munin the necessary portion of care, and shrinks my `extraConfig` to 1/8th. Did not test `extraAutoPlugins`, though.

ToxicFrog committed:
nixos/munin: add types to Munin options

Some options were missing their types.

ToxicFrog committed:
nixos/munin: enable munin_update and disable munin_stats

munin_update relies on a stats file that exists, but isn't found in the
default location on NixOS; the appropriate plugin configuration is
added.

munin_stats relies on munin-cron writing a logfile, which the NixOS
build of munin does not. (This is probably fixable in the munin package,
but I don't have time to dig into that right now.)

ToxicFrog commented:
I've updated tests/munin.nix to use the new disabledPlugins setting, and it seems to work.

florianjacob commented:
@GrahamcOfBorg test munin

florianjacob commented:
@GrahamcOfBorg build nixosTests.munin

florianjacob commented:
@GrahamcOfBorg build nixosTests.munin

ToxicFrog committed:
munin: 2.0.37 -> 2.0.43

Also creates the RELEASE file in preBuild so that munin knows its own
version number at runtime.

ToxicFrog commented:
Made a small tweak to the CSS commit (so that the example code doesn't look hideous on warning/critical graphs) and added a new commit to bump the Munin version to the latest stable release (and fix a small bug in the build).

FRidh commented:
@GrahamcOfBorg build nixosTests.munin

ToxicFrog commented:
Is there anything else I need to do here (sync and resolve conflicts?) or is it just waiting for a maintainer to merge it?

florianjacob commented:
@FRidh do you have time for a review, or can recommend someone else?

Infinisil commented:
Alright I think it's fine to stay. When my other comment is addressed and the merge conflict fixed this is fine to merge.

ToxicFrog commented on nixos/modules/services/monitoring/munin.nix:
```diff
@@ -114,6 +116,17 @@ let
+    cd "$out"
+    chmod -R u+w .
+    cat "$cssPath" >> style.css
+    cat "$cssPath" >> style-new.css
```
`cp` won't work here because the goal is to append it to the existing CSS files, since it's not meant to completely replace the built-in CSS, just (potentially) override selected rules. I've replaced `passAsFile` and `cat` with appropriate use of `echo >>` instead.

ToxicFrog commented:
Rebased against latest master, conflicts resolved.

domenkozar commented:
@GrahamcOfBorg build tests.munin

Infinisil commented:
@GrahamcOfBorg test munin
